README
aconnectgui is a FLTK based frontend for aconnect. It is written
directly on top of the alsamixer source, leaving the original source
intact, only adding a couple of ifdefs, and some calls to the gui
part, so it provides exactly the same functionality, but with a 
graphical userinterface.



ChangeLog
0.9rc1-2  - Applied patch from tvignaud@mandrakesoft.com to fix compilation
            on Mandrake (Fl/FL includes)



src/Port.cxx
#include "Port.hxx"
#include "Connector.hxx"
#include <FL/Fl_Box.H>

Port::Port(int ix,int iy,int iw,int ih,char* il,int portId,ConnectorType t)
:Fl_Group(ix,iy,iw,ih)
{
	mLabelBox = new Fl_Box(
		x(),y()+CONNECTOR_OFFSET,
		w()-8,h()-CONNECTOR_OFFSET-PORT_EXTRA
	);
	mLabelBox->box(FL_FLAT_BOX);
	mLabelBox->label(il);
	mLabelBox->labelsize(10);
	mLabelBox->align(FL_ALIGN_INSIDE|FL_ALIGN_RIGHT);

	mPortId = portId;

	int cy=y();
	
	mInput = 0;
	mOutput = 0;
	
	if (t&kInput) {
		add(mInput = new Connector(x()+w()-5,cy,kInput));

		cy+=CONNECTOR_DISTANCE;				
	}
	if (t&kOutput) {
		add(mOutput = new Connector(x()+w()-5,cy,kOutput));
	}
	
	resizable(0);
}

Connector* Port::FindConnector(int ix,int iy)
{
	if (ix>=x() && ix<=x()+w() && iy>=y() && iy<=y()+h()) {
		if (mInput && mInput->Contains(ix,iy)) {
			return mInput;
		}
		if (mOutput && mOutput->Contains(ix,iy)) {
			return mOutput;
		}
	}
	return 0;
}	


src/Client.cxx
#include "MyWindow.hxx"
#include "Client.hxx"
#include "Connection.hxx"
#include "Port.hxx"
#include "Layout.hxx"
#include <FL/Fl.H>

Client::Client(int X,int Y,int W,int H,char* L,int C)
:FrameGroup(X,Y,W,H,L)
{
	labelsize(10);
	mPorts = new Fl_Group(X+2,Y+14,W-7,0);
	mPorts->end();
	mPorts->resizable(0);
	end();
	resizable(0);
	mClientId = C;
}

Port* Client::FindPort(int clientId,int portId)
{
	if (mClientId!=clientId) return 0;
	
	int n = mPorts->children();
	for (int i=0;i<n;i++) {
		Port* p = (Port*) (mPorts->child(i));
		if (p->PortId()==portId) {
			return p;
		}
	}
	return 0;
}

Connector* Client::FindInput(int clientId,int portId)
{
	if (mClientId!=clientId) return 0;
	
	int n = mPorts->children();
	for (int i=0;i<n;i++) {
		Port* p = (Port*) (mPorts->child(i));
		if (p->PortId()==portId && p->Input()) {
			return p->Input();
		}
	}
	return 0;
}

Connector* Client::FindOutput(int clientId,int portId)
{
	if (mClientId!=clientId) return 0;
	
	int n = mPorts->children();
	for (int i=0;i<n;i++) {
		Port* p = (Port*) (mPorts->child(i));
		if (p->PortId()==portId && p->Output()) {
			return p->Output();
		}
	}
	return 0;
}

Connector* Client::FindConnector(int ix,int iy)
{
	if (ix>=x() && ix<=x()+w() && iy>=y() && iy<=y()+h()) {
		int n = mPorts->children();
		for (int i=0;i<n;i++) {
			Port* p = (Port*) (mPorts->child(i));
			Connector* c = p->FindConnector(ix,iy);
			if (c) return c;
		}
	}
	return 0;
}

Clients::Clients(int X,int Y,int W,int H):Fl_Group(X,Y,W,H)
{ 
	resizable(0);	
	end(); 
}

Port*      Clients::FindPort(int clientId,int portId)
{
	int n = children();
	for (int i=0;i<n;i++) {
		Client* c = (Client*) (child(i));
		Port* p = c->FindPort(clientId,portId);
		if (p) return p;
	}
	return 0;
}

Connector* Clients::FindInput(int clientId,int portId)
{
	int n = children();
	for (int i=0;i<n;i++) {
		Client* c = (Client*) (child(i));
		Connector* r = c->FindInput(clientId,portId);
		if (r) return r;
	}
	return 0;
}

Connector* Clients::FindOutput(int clientId,int portId)
{
	int n = children();
	for (int i=0;i<n;i++) {
		Client* c = (Client*) (child(i));
		Connector* r = c->FindOutput(clientId,portId);
		if (r) return r;
	}
	return 0;
}

Client*    Clients::FindClient(int clientId)
{
	int n = children();
	for (int i=0;i<n;i++) {
		Client* c = (Client*) (child(i));
		if (c->ClientId()==clientId) return c;
	}
	return 0;
}

Connector* Clients::FindConnector(int ix,int iy)
{
	int n = children();
	for (int i=0;i<n;i++) {
		Client* c = (Client*) (child(i));
		Connector* con = c->FindConnector(ix,iy);
		if (con) return con;
	}
	return 0;
}

Port* Client::AddPort(char* il,int portId,ConnectorType t)
{
	Port* port;
	int ph=0;
	if (t&kInput) {
		ph+=CONNECTOR_DISTANCE;				
	}
	if (t&kOutput) {
		ph+=CONNECTOR_DISTANCE;
	}
	ph+=PORT_EXTRA;

	port = new Port(mPorts->x(),mPorts->y()+mPorts->h(),mPorts->w(),ph,il,portId,t);
	mPorts->size(mPorts->w(),port->y()+port->h()-mPorts->y());
	mPorts->add(port);
	size(w(),mPorts->y()+mPorts->h()-y()+4);
	mPorts->color(FL_RED);
	mPorts->redraw();
	return port;
}

int Clients::handle(int e)
{
	static Connector* a,*b,*prev;
	static Connection* c=0;

	MyWindow* patchbay = ((MyWindow*)parent());
	Connections* connections = patchbay->GetConnections();

	if (e==FL_PUSH && patchbay->IsConnecting()) {
		a = 0; b = 0; c = 0; prev = 0;
		a = FindConnector(Fl::event_x(),Fl::event_y());
		if (a) {
			a->Select();
			a->redraw();
		}
	}
	if (e==FL_DRAG && a) {
		b = FindConnector(Fl::event_x(),Fl::event_y());
		if (prev!=b && a!=b) {
			if (prev) {
				prev->Unselect();
				prev->redraw();
			}
			if (b) {
				if (a!=b) {
					if (!connections->Exists(a,b)) {
						if (patchbay->IsLegal(a,b)) {
							c= new Connection(
								connections->x(),connections->y(),
								connections->w(),connections->h()
							);
							b->Select();
							b->redraw();
							c->Set(a,b);
							connections->add(c);
							connections->redraw();
						}
					}
				}
			}else{
				if (c) {
					connections->remove(c);
					delete c;
					c = 0;
					connections->redraw();
				}
			}
			prev=b;
		}
	}
	if (e==FL_RELEASE) {
		if (c) {
			connections->remove(c);
			delete c;
			c = 0;
		}
		if (a && b) {
			bool flag = 0;
			if (a->Type()==kOutput)
			{
				flag = patchbay->HandleConnect(a,b);
			}else{
				flag = patchbay->HandleConnect(b,a);
			}
			if (!flag)
			{
				a->Unselect();
				a->redraw();
				b->Unselect();
				b->redraw();
			}
		}else{
			if (a) {
				a->Unselect();
				a->redraw();
			}
		}
	}
	return 1;
}


src/Window.cxx
#include "Window.hxx"
#include "Connector.hxx"
#include "Port.hxx"
#include <FL/Fl.H>
#include <FL/fl_ask.H>

#include "images/scicors.xpm"
#include "images/connect.xpm"
#include "images/question.xpm"
#include "images/alsalogo.xpm"

static Fl_Pixmap pixmap_scicors(scicors_xpm);
static Fl_Pixmap pixmap_connect(connect_xpm);
static Fl_Pixmap pixmap_question(question_xpm);
static Fl_Pixmap pixmap_alsalogo(alsalogo_xpm);

int aconnect_main(int argc,char** argv);

void Window::Connect(int client,int port,int clientB,int portB)
{
	Connector *a = 0,*b = 0;
	
	a = mClients->FindOutput(client,port);
	b = mClients->FindInput(clientB,portB);
//	if (a==0) printf("%d %d does not have an output\n",client,port);
//	if (b==0) printf("%d %d does not have an input\n",clientB,portB);
	if (a!=0 && b!=0) {
		Connect(a,b);
	}
}		

void Window::Disconnect(
	int fromClientId,int fromPortId,int toClientId,int toPortId)
{
	Connection* c = mConnections->Find(
		fromClientId,fromPortId,toClientId,toPortId);
/*
	if (!c)
		printf("could not find connection between %d:%d %d:%d\n",
			client,port,clientB,portB);
*/
	if (c)
	{
		Disconnect(c);
	}
}

void Window::Disconnect(Connection* c)
{
	mConnections->remove(c);
	mConnections->redraw();
	c->From()->Disconnect();
	c->To()->Disconnect();
	c->From()->redraw();
	c->To()->redraw();
	delete c;
	mConnections->Unclutter();
	mConnections->redraw();
}

void Window::Connect(Connector* a,Connector* b)
{
	a->Unselect();
	b->Unselect();

	Connection* c= new Connection(
		mConnections->x(),mConnections->y(),
		mConnections->w(),mConnections->h()
	);

	c->Set(a,b);

	a->Connect();
	b->Connect();

	a->redraw();
	b->redraw();

	mConnections->add(c);
	mConnections->Unclutter();
}

int Window::IsLegal(Connector* a,Connector* b)
{
	return a->Type()!=b->Type();
}

Client* Window::AddClient(snd_seq_client_info_t* cinfo)
{
	// some of this code should be in Window
	Client* c= mClients->FindClient(snd_seq_client_info_get_client(cinfo));
	
	if (c) {
		mCurClient = c;
	}else{
		mCurClient = new Client(
			mClients->x(),mClients->y()+mClients->h(),mClients->w(),20,
			strdup(snd_seq_client_info_get_name(cinfo)),
			snd_seq_client_info_get_client(cinfo)
		);
		int newh = mCurClient->y()+mCurClient->h()-mClients->y();
		
		mClients->size(mClients->w(),newh);
		mConnections->size(mConnections->w(),newh);
		mClients->add(mCurClient);
		size(w(),newh+33);
	}
	return mCurClient;
}
Port* Window::AddPort(snd_seq_port_info_t* pinfo)
{
	// some of this code should be in Window
	int dy = mCurClient->y()+mCurClient->h();
	mCurPort = mCurClient->AddPort(
		strdup(snd_seq_port_info_get_name(pinfo)),
	  snd_seq_port_info_get_port(pinfo),
		mCurType);
	dy = mCurClient->y()+mCurClient->h()-dy;
	for (int i = mClients->find(mCurClient)+1;i<mClients->children();i++)
	{
		mClients->child(i)->position(mClients->x(),mClients->child(i)->y()+dy);
	}
	int newh = mClients->child(mClients->children()-1)->y()+
		mClients->child(mClients->children()-1)->h()-
		mClients->y();
	mClients->size(mClients->w(),newh);
	mConnections->size(mConnections->w(),newh);
	size(w(),newh+33);
	return mCurPort;
}

bool Window::HandleConnect(Connector* a,Connector* b)
{
	int argc;
	char* argv[16];
	char from[16];
	char to[16];
	
	sprintf(from,"%d:%d",a->Owner()->ClientId(),a->Owner()->PortId());
	sprintf(to,"%d:%d",b->Owner()->ClientId(),b->Owner()->PortId());
	
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = from;
	argv[argc++] = to;

	return (aconnect_main(argc,argv)==0);
}

bool Window::HandleDisconnect(Connection* c)
{
	int argc;
	char* argv[16];
	char from[16];
	char to[16];
	
	Connector* a = c->From();
	Connector* b = c->To();
	
	sprintf(from,"%d:%d",a->Owner()->ClientId(),a->Owner()->PortId());
	sprintf(to,"%d:%d",b->Owner()->ClientId(),b->Owner()->PortId());
	
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-d";
	argv[argc++] = from;
	argv[argc++] = to;
	
	return (aconnect_main(argc,argv)==0);
}

Window::Window():Fl_Window(300,33,"ALSA Sequencer")
{
	mClients = new Clients(0,33,140,0);
	mConnections = new Connections(140,33,160,0);
	int cx = 0;
	
	Fl_Group* g = new Fl_Group(0,0,300,24);
	{
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_scicors.label(b);
		mCutButton = b;
		cx+=24;
	}
	{
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_connect.label(b);
		mConnectButton = b;
		cx+=24;
	}
	if (0) {
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_question.label(b);
		mInfoButton = b;
		cx+=24;
	}
	{
		Fl_Box* b = new Fl_Box(cx,0,300-cx,24);
		b->box(FL_THIN_UP_BOX);
		b->align(FL_ALIGN_INSIDE|FL_ALIGN_RIGHT);
		pixmap_alsalogo.label(b);
		cx+=24;
	}
	{
		Fl_Box* b = new Fl_Box(0,24,300,1);
		b->color(FL_WHITE);
		b->box(FL_FLAT_BOX);
	}
	g->end();

	mCurClient = 0;
	mCurPort = 0;
	mCurType = kNone;
	mHandle = 0;
	if (snd_seq_open(&mHandle,"hw",SND_SEQ_OPEN_INPUT,0)<0)
	{
		fl_alert("Error opening sequencer\n");
		exit(-1);
	}
	mClientId = snd_seq_client_id(mHandle);
	if(mClientId<0)
	{
		fl_alert("Error opening sequencer\n");
		exit(-1);
	}
	snd_seq_set_client_name(mHandle,"aconnectgui");
	unsigned int caps = SND_SEQ_PORT_CAP_WRITE;
	int port =  snd_seq_create_simple_port(mHandle, "Notify", caps,
		SND_SEQ_PORT_TYPE_MIDI_GENERIC |
		SND_SEQ_PORT_TYPE_APPLICATION);
		
	if (port<0)
	{
		fl_alert("Error creating port\n");
		exit(-1);
	}
	int err;
	if ((err=snd_seq_connect_from(mHandle,port,0,1))<0)
	{
		fl_alert("Error subscribing to System::Announce port [%d]\n",err);
		exit(-1);
	}

	mPollMax = snd_seq_poll_descriptors_count(mHandle,POLLIN);
	mPollFds = (struct pollfd *) calloc(mPollMax, sizeof(struct pollfd));

	snd_seq_nonblock(mHandle, 1);
}

Window::~Window()
{
	snd_seq_close(mHandle);
}


void Window::Timeout(void)
{
	int count;
	snd_seq_event_t *ev;

	do
	{
		snd_seq_event_input(mHandle, &ev);
		if (ev)
		{
			snd_seq_free_event(ev);
			if (ev->type==SND_SEQ_EVENT_PORT_SUBSCRIBED)
			{
				if (
					ev->data.connect.sender.client!=mClientId &&
					ev->data.connect.dest.client!=mClientId)
				{
					Connect(
						ev->data.connect.sender.client,
						ev->data.connect.sender.port,
						ev->data.connect.dest.client,
						ev->data.connect.dest.port);
					mConnections->redraw();
				}
			}
			if (ev->type==SND_SEQ_EVENT_PORT_UNSUBSCRIBED)
			{
				if (
					ev->data.connect.sender.client!=mClientId &&
					ev->data.connect.dest.client!=mClientId)
				{
					Disconnect(
						ev->data.connect.sender.client,
						ev->data.connect.sender.port,
						ev->data.connect.dest.client,
						ev->data.connect.dest.port);
					mConnections->redraw();
				}
			}
			if (ev->type==SND_SEQ_EVENT_CLIENT_START)
			{
				if (ev->data.addr.client != snd_seq_client_id(mHandle))
				{
					snd_seq_client_info_t *cinfo;
					snd_seq_port_info_t *pinfo;
					snd_seq_client_info_alloca(&cinfo);
					snd_seq_port_info_alloca(&pinfo);
					if (snd_seq_get_any_client_info(mHandle, ev->data.addr.client, cinfo) >= 0) {
						AddClient(cinfo);
						/* reset query info */
						count = 0;
						if (snd_seq_get_any_port_info(mHandle, 
							ev->data.addr.client,
							 ev->data.addr.port,
							pinfo) >= 0) {
							if (snd_seq_port_info_get_capability(pinfo) & (SND_SEQ_PORT_CAP_WRITE|SND_SEQ_PORT_CAP_SUBS_WRITE))
							{
								mCurType = kInput;
								AddPort(pinfo);
							}
							if (snd_seq_port_info_get_capability(pinfo) & (SND_SEQ_PORT_CAP_READ|SND_SEQ_PORT_CAP_SUBS_READ))
							{
								mCurType = kOutput;
								AddPort(pinfo);
							}
						}
					}
					fflush(stdout);
				}	
			}
			if (ev->type==SND_SEQ_EVENT_CLIENT_EXIT)
			{
				Client* c= mClients->FindClient(ev->data.addr.client);
				int newh = mClients->h()-c->h();
				mConnections->RemoveClient(ev->data.addr.client);
				mConnections->size(mConnections->w(),newh);
				mClients->remove(c);
				mClients->size(mClients->w(),newh);
				size(w(),newh+33);
			}
			snd_seq_free_event(ev);
		}
	} while (ev);

	Fl::add_timeout(0.1,Window::TimeoutStatic,this);
}

Window* patchbay = 0;

main()
{
	int argc;
	char* argv[16];
	
	Fl::get_system_colors();
	
	patchbay = new Window;

	/* NB the concept of input of aconnect and the  classes is reverse...
	*/

	patchbay->SetCurType(kOutput);
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-i";
	argv[argc++] = "-l";
	aconnect_main(argc,argv);

	patchbay->SetCurType(kInput);
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-o";
	argv[argc++] = "-l";
	aconnect_main(argc,argv);

	patchbay->show();

	Fl::add_timeout(0.1,Window::TimeoutStatic,patchbay);

	Fl::run();
}


src/Connector.cxx
#include "Connection.hxx"
#include "Connector.hxx"

#include <FL/fl_draw.H>

Connector::Connector(int ix,int iy,ConnectorType t)
:Fl_Widget(ix,iy,8,CONNECTOR_DISTANCE) {
	mType=t;
	mConnectedCount=0;
	mIsSelected=false;
}

void Connector::draw()
{
	int cx=x(),cy=y()+CONNECTOR_OFFSET;
	if (mType==kOutput) {
		fl_color(FL_BLACK);
		fl_yxline(cx,cy,cy+8);
		fl_line(cx+1,cy,cx+5,cy+4);
		fl_line(cx+5,cy+4,cx+1,cy+8);
		if (IsSelected()) 
			fl_color(FL_DARK1);
		else
			if (Connected()) 
				fl_color(FL_BLACK);			
			else 
				fl_color(FL_WHITE);
		for (int k=0;k<4;k++) {
			fl_xyline(cx+1,cy+k+1,cx+1+k);
			fl_xyline(cx+1,cy+7-k,cx+1+k);
		}			
	}else{
		fl_color(FL_BLACK);
		fl_yxline(cx+5,cy,cy+8);
		fl_line(cx+4,cy,cx,cy+4);
		fl_line(cx,cy+4,cx+4,cy+8);
		if (IsSelected()) 
			fl_color(FL_DARK1);
		else
			if (Connected()) 
				fl_color(FL_BLACK);			
			else 
				fl_color(FL_WHITE);
		for (int k=0;k<4;k++) {
			fl_xyline(cx+4-k,cy+k+1,cx+4);
			fl_xyline(cx+4-k,cy+7-k,cx+4);
		}					
	}
}

bool Connector::Contains(int ix,int iy)
{
	return (ix>=x() && ix<=x()+w() && iy>=y() && iy<=y()+h());
}



src/MyWindow.cxx
#include "MyWindow.hxx"
#include "Connector.hxx"
#include "Port.hxx"
#include <FL/Fl.H>
#include <FL/fl_ask.H>

#include "images/scicors.xpm"
#include "images/connect.xpm"
#include "images/question.xpm"
#include "images/alsalogo.xpm"

static Fl_Pixmap pixmap_scicors(scicors_xpm);
static Fl_Pixmap pixmap_connect(connect_xpm);
static Fl_Pixmap pixmap_question(question_xpm);
static Fl_Pixmap pixmap_alsalogo(alsalogo_xpm);

int aconnect_main(int argc,char** argv);

void MyWindow::Connect(int client,int port,int clientB,int portB)
{
	Connector *a = 0,*b = 0;
	
	a = mClients->FindOutput(client,port);
	b = mClients->FindInput(clientB,portB);
//	if (a==0) printf("%d %d does not have an output\n",client,port);
//	if (b==0) printf("%d %d does not have an input\n",clientB,portB);
	if (a!=0 && b!=0) {
		Connect(a,b);
	}
}		

void MyWindow::Disconnect(
	int fromClientId,int fromPortId,int toClientId,int toPortId)
{
	Connection* c = mConnections->Find(
		fromClientId,fromPortId,toClientId,toPortId);
/*
	if (!c)
		printf("could not find connection between %d:%d %d:%d\n",
			client,port,clientB,portB);
*/
	if (c)
	{
		Disconnect(c);
	}
}

void MyWindow::Disconnect(Connection* c)
{
	mConnections->remove(c);
	mConnections->redraw();
	c->From()->Disconnect();
	c->To()->Disconnect();
	c->From()->redraw();
	c->To()->redraw();
	delete c;
	mConnections->Unclutter();
	mConnections->redraw();
}

void MyWindow::Connect(Connector* a,Connector* b)
{
	a->Unselect();
	b->Unselect();

	Connection* c= new Connection(
		mConnections->x(),mConnections->y(),
		mConnections->w(),mConnections->h()
	);

	c->Set(a,b);

	a->Connect();
	b->Connect();

	a->redraw();
	b->redraw();

	mConnections->add(c);
	mConnections->Unclutter();
}

int MyWindow::IsLegal(Connector* a,Connector* b)
{
	return a->Type()!=b->Type();
}

Client* MyWindow::AddClient(snd_seq_client_info_t* cinfo)
{
	// some of this code should be in MyWindow
	Client* c= mClients->FindClient(snd_seq_client_info_get_client(cinfo));
	
	if (c) {
		mCurClient = c;
	}else{
		mCurClient = new Client(
			mClients->x(),mClients->y()+mClients->h(),mClients->w(),20,
			strdup(snd_seq_client_info_get_name(cinfo)),
			snd_seq_client_info_get_client(cinfo)
		);
		int newh = mCurClient->y()+mCurClient->h()-mClients->y();
		
		mClients->size(mClients->w(),newh);
		mConnections->size(mConnections->w(),newh);
		mClients->add(mCurClient);
		size(w(),newh+33);
	}
	return mCurClient;
}
Port* MyWindow::AddPort(snd_seq_port_info_t* pinfo)
{
	// some of this code should be in MyWindow
	int dy = mCurClient->y()+mCurClient->h();
	mCurPort = mCurClient->AddPort(
		strdup(snd_seq_port_info_get_name(pinfo)),
	  snd_seq_port_info_get_port(pinfo),
		mCurType);
	dy = mCurClient->y()+mCurClient->h()-dy;
	for (int i = mClients->find(mCurClient)+1;i<mClients->children();i++)
	{
		mClients->child(i)->position(mClients->x(),mClients->child(i)->y()+dy);
	}
	int newh = mClients->child(mClients->children()-1)->y()+
		mClients->child(mClients->children()-1)->h()-
		mClients->y();
	mClients->size(mClients->w(),newh);
	mConnections->size(mConnections->w(),newh);
	size(w(),newh+33);
	return mCurPort;
}

bool MyWindow::HandleConnect(Connector* a,Connector* b)
{
	int argc;
	char* argv[16];
	char from[16];
	char to[16];
	
	sprintf(from,"%d:%d",a->Owner()->ClientId(),a->Owner()->PortId());
	sprintf(to,"%d:%d",b->Owner()->ClientId(),b->Owner()->PortId());
	
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = from;
	argv[argc++] = to;

	return (aconnect_main(argc,argv)==0);
}

bool MyWindow::HandleDisconnect(Connection* c)
{
	int argc;
	char* argv[16];
	char from[16];
	char to[16];
	
	Connector* a = c->From();
	Connector* b = c->To();
	
	sprintf(from,"%d:%d",a->Owner()->ClientId(),a->Owner()->PortId());
	sprintf(to,"%d:%d",b->Owner()->ClientId(),b->Owner()->PortId());
	
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-d";
	argv[argc++] = from;
	argv[argc++] = to;
	
	return (aconnect_main(argc,argv)==0);
}

MyWindow::MyWindow():Fl_Window(300,33,"ALSA Sequencer")
{
	mClients = new Clients(0,33,140,0);
	mConnections = new Connections(140,33,160,0);
	int cx = 0;
	
	Fl_Group* g = new Fl_Group(0,0,300,24);
	{
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_scicors.label(b);
		mCutButton = b;
		cx+=24;
	}
	{
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_connect.label(b);
		mConnectButton = b;
		cx+=24;
	}
	if (0) {
		Fl_Radio_Button* b = new Fl_Radio_Button(cx,0,24,24);
		b->box(FL_THIN_UP_BOX);
		b->down_box(FL_FLAT_BOX);
		pixmap_question.label(b);
		mInfoButton = b;
		cx+=24;
	}
	{
		Fl_Box* b = new Fl_Box(cx,0,300-cx,24);
		b->box(FL_THIN_UP_BOX);
		b->align(FL_ALIGN_INSIDE|FL_ALIGN_RIGHT);
		pixmap_alsalogo.label(b);
		cx+=24;
	}
	{
		Fl_Box* b = new Fl_Box(0,24,300,1);
		b->color(FL_WHITE);
		b->box(FL_FLAT_BOX);
	}
	g->end();

	mCurClient = 0;
	mCurPort = 0;
	mCurType = kNone;
	mHandle = 0;
	if (snd_seq_open(&mHandle,"hw",SND_SEQ_OPEN_INPUT,0)<0)
	{
		fl_alert("Error opening sequencer\n");
		exit(-1);
	}
	mClientId = snd_seq_client_id(mHandle);
	if(mClientId<0)
	{
		fl_alert("Error opening sequencer\n");
		exit(-1);
	}
	snd_seq_set_client_name(mHandle,"aconnectgui");
	unsigned int caps = SND_SEQ_PORT_CAP_WRITE;
	int port =  snd_seq_create_simple_port(mHandle, "Notify", caps,
		SND_SEQ_PORT_TYPE_MIDI_GENERIC |
		SND_SEQ_PORT_TYPE_APPLICATION);
		
	if (port<0)
	{
		fl_alert("Error creating port\n");
		exit(-1);
	}
	int err;
	if ((err=snd_seq_connect_from(mHandle,port,0,1))<0)
	{
		fl_alert("Error subscribing to System::Announce port [%d]\n",err);
		exit(-1);
	}

	mPollMax = snd_seq_poll_descriptors_count(mHandle,POLLIN);
	mPollFds = (struct pollfd *) calloc(mPollMax, sizeof(struct pollfd));

	snd_seq_nonblock(mHandle, 1);
}

MyWindow::~MyWindow()
{
	snd_seq_close(mHandle);
}


void MyWindow::Timeout(void)
{
	int count;
	snd_seq_event_t *ev;

	do
	{
		snd_seq_event_input(mHandle, &ev);
		if (ev)
		{
			snd_seq_free_event(ev);
			if (ev->type==SND_SEQ_EVENT_PORT_SUBSCRIBED)
			{
				if (
					ev->data.connect.sender.client!=mClientId &&
					ev->data.connect.dest.client!=mClientId)
				{
					Connect(
						ev->data.connect.sender.client,
						ev->data.connect.sender.port,
						ev->data.connect.dest.client,
						ev->data.connect.dest.port);
					mConnections->redraw();
				}
			}
			if (ev->type==SND_SEQ_EVENT_PORT_UNSUBSCRIBED)
			{
				if (
					ev->data.connect.sender.client!=mClientId &&
					ev->data.connect.dest.client!=mClientId)
				{
					Disconnect(
						ev->data.connect.sender.client,
						ev->data.connect.sender.port,
						ev->data.connect.dest.client,
						ev->data.connect.dest.port);
					mConnections->redraw();
				}
			}
			if (ev->type==SND_SEQ_EVENT_CLIENT_START)
			{
				if (ev->data.addr.client != snd_seq_client_id(mHandle))
				{
					snd_seq_client_info_t *cinfo;
					snd_seq_port_info_t *pinfo;
					snd_seq_client_info_alloca(&cinfo);
					snd_seq_port_info_alloca(&pinfo);
					if (snd_seq_get_any_client_info(mHandle, ev->data.addr.client, cinfo) >= 0) {
						AddClient(cinfo);
						/* reset query info */
						count = 0;
						if (snd_seq_get_any_port_info(mHandle, 
							ev->data.addr.client,
							 ev->data.addr.port,
							pinfo) >= 0) {
							if (snd_seq_port_info_get_capability(pinfo) & (SND_SEQ_PORT_CAP_WRITE|SND_SEQ_PORT_CAP_SUBS_WRITE))
							{
								mCurType = kInput;
								AddPort(pinfo);
							}
							if (snd_seq_port_info_get_capability(pinfo) & (SND_SEQ_PORT_CAP_READ|SND_SEQ_PORT_CAP_SUBS_READ))
							{
								mCurType = kOutput;
								AddPort(pinfo);
							}
						}
					}
					fflush(stdout);
				}	
			}
			if (ev->type==SND_SEQ_EVENT_CLIENT_EXIT)
			{
				Client* c= mClients->FindClient(ev->data.addr.client);
				int newh = mClients->h()-c->h();
				mConnections->RemoveClient(ev->data.addr.client);
				mConnections->size(mConnections->w(),newh);
				mClients->remove(c);
				mClients->size(mClients->w(),newh);
				size(w(),newh+33);
			}
			snd_seq_free_event(ev);
		}
	} while (ev);

	Fl::add_timeout(0.1,MyWindow::TimeoutStatic,this);
}

MyWindow* patchbay = 0;

main()
{
	int argc;
	char* argv[16];
	
	Fl::get_system_colors();
	
	patchbay = new MyWindow;

	/* NB the concept of input of aconnect and the  classes is reverse...
	*/

	patchbay->SetCurType(kOutput);
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-i";
	argv[argc++] = "-l";
	aconnect_main(argc,argv);

	patchbay->SetCurType(kInput);
	argc = 0;
	argv[argc++] = "aconnect";
	argv[argc++] = "-o";
	argv[argc++] = "-l";
	aconnect_main(argc,argv);

	patchbay->show();

	Fl::add_timeout(0.1,MyWindow::TimeoutStatic,patchbay);

	Fl::run();
}
